### PR TITLE
borrowings_service implemented

### PR DESCRIPTION
**Pull request by DimaYankovskiy**

### CODE REVIEW

File: borrowings_service/models.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+from book_service.models import Book
+from library_service import settings
+
+
+class Borrowing(models.Model):
+    borrow_date = models.DateField(default=datetime.now)
+    expected_return_date = models.DateField()
+    actual_return_date = models.DateField(null=True, blank=True)
+    book = models.ForeignKey(Book, on_delete=models.CASCADE)
+    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
```
Comment (Yuruch) [inline borrowings_service/models.py:12]: Borrowing:
Borrow date: date
Expected return date: date
Actual return date: date
Book id: int
User id: int

i guess here should be book id and user id not ForeignKey)

File: borrowings_service/models.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
+from datetime import datetime
+
+from django.db import models
+
+from book_service.models import Book
+from library_service import settings
```
Comment (Yuruch) [inline borrowings_service/models.py]: settings should be imported only from django.conf

File: borrowings_service/serializers.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        model = Borrowing
+        fields = (
+            "expected_return_date",
+            "book",
+        )
+
+    def validate(self, attrs):
+        if attrs["book"].inventory <= 0:
+            raise serializers.ValidationError("The selected book is not available.")
+        return attrs
```
Comment (Yuruch) [inline borrowings_service/serializers.py:37]: Maybe we can just filter books that are not available.
Also expected_return_date should be validated ( > than now)

File: borrowings_service/serializers.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+class BorrowingReturnSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Borrowing
+        fields = ("actual_return_date",)
+
+    def validate_actual_return_date(self, value):
+        if value and value < self.instance.borrow_date:
+            raise serializers.ValidationError("Return date cannot be before borrow date.")
+        return value
```
Comment (Yuruch) [inline borrowings_service/serializers.py]: This validation will never raise ValidationError

File: borrowings_service/views.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+from borrowings_service.models import Borrowing
+from borrowings_service.serializers import (
+    BorrowingReadSerializer,
+    BorrowingCreateSerializer,
+    BorrowingReturnSerializer,
+)
+
+
+class BorrowingViewSet(viewsets.ModelViewSet):
+    queryset = Borrowing.objects.all()
```
Comment (Yuruch) [inline borrowings_service/views.py:18]: You don't need queryset  if u redefine get_queryset method

File: borrowings_service/views.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+
+        return queryset
+
+    def perform_create(self, serializer):
+        book = serializer.validated_data["book"]
+        if book.inventory <= 0:
+            raise ValidationError("The selected book is not available.")
+
+        book.inventory -= 1
+        book.save()
```
Comment (Yuruch) [inline borrowings_service/views.py:53]: we also will not need this validation if we just filter books

File: borrowings_service/views.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+        borrowing.actual_return_date = datetime.now().date()
+        serializer = self.get_serializer(borrowing, data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        borrowing.book.inventory += 1
+        borrowing.book.save()
+
+        return Response(serializer.data)
```
Comment (Yuruch) [inline borrowings_service/views.py]: add status code here

**Review by Yuruch — CHANGES_REQUESTED** [review]
Thank you for your hard work, but please make some changes :)

File: borrowings_service/models.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class Borrowing(models.Model):
+    borrow_date = models.DateField(default=datetime.now)
```
Comment (Yuruch) [inline borrowings_service/models.py]: we can use auto_now_add=True here

File: borrowings_service/admin.py [hunk 9]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
+
+# Register your models here.
```
Comment (necobra) [inline borrowings_service/admin.py]: maybe just register the model

**Review by Yuruch — APPROVED** [review]
Good job! Thanks for your effort

File: borrowings_service/views.py [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+
+        borrowing.actual_return_date = datetime.now().date()
+        serializer = self.get_serializer(borrowing, data=request.data)
+        serializer.is_valid(raise_exception=True)
+        serializer.save()
+
+        borrowing.book.inventory += 1
+        borrowing.book.save()
+
+        return Response(serializer.data, status=status.HTTP_200_OK)
```
Comment (necobra) [inline borrowings_service/views.py:78]: I see no validation that the same user who took it is returning it. Did you check?

**Review by necobra — APPROVED** [review]

**Pull request merged**